src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -156,10 +156,10 @@ public void CompleteStream(bool errored)
                 // connection's flow-control window.
                 _inputFlowControl.Abort();
 
-                // We only want to reuse a stream that has completely finished writing.
-                // This is to prevent the situation where Http2OutputProducer.ProcessDataWrites
-                // is still running in the background.
-                CanReuse = !_keepAlive && HasResponseCompleted;
+                // We only want to reuse a stream that has completely finished writing
+                // and the request wasn't aborted. This is to ensure
+                // Http2OutputProducer.ProcessDataWrites is in the correct state to be reused.
+                CanReuse = !_keepAlive && HasResponseCompleted && !_connectionAborted;
```

Comment:
To make sure, _connectionAborted, _keepAlive, and HasResponseComplete are not updated in a lock at all or expect to be check while holding a lock?

Reply:
_connectionAborted might be. I don't know whether it should be inside a lock.